### PR TITLE
Add information_schema.user_mapping_options

### PR DESCRIPTION
**Pull request by jeeminso**
## Summary of the changes / Why this improves CrateDB
Exposes user mapping options main logic to check are 1) converting to/from xcontent and 2) masking option values (to protect the pw).

## Checklist

 - [x] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by jeeminso** [comment]
Follow up: reject unsupported user mapping options, currently only should accept `user` and `password`.

File: server/src/main/java/io/crate/metadata/information/InformationSchemaTableDefinitions.java [hunk 1]
```diff
@@ -129,6 +129,18 @@ public InformationSchemaTableDefinitions(Roles roles,
...
         ));
+        tableDefinitions.put(UserMappingOptionsTableInfo.IDENT, new StaticTableDefinition<>(
+            informationSchemaIterables::userMappingOptions,
+            (user, t) -> roles.hasPrivilege(user, Permission.AL, Securable.CLUSTER, null),
+            (user, t) -> {
+                if ((user.name().equals(t.userName()) || user.isSuperUser()) == false) {
+                    return new UserMappingOptionsTableInfo.UserMappingOptions(
+                        t.userName(), t.serverName(), t.optionName(), null);
+                }
+                return t;
```
Comment (jeeminso) [inline server/src/main/java/io/crate/metadata/information/InformationSchemaTableDefinitions.java:140]: currently all option values are masked not just pw.

**Comment by jeeminso** [comment]
Follow up:
```
postgres=# CREATE USER MAPPING FOR user1 SERVER foo OPTIONS (user 'bob', password 'secret');
CREATE USER MAPPING
postgres=# drop user user1;
ERROR:  role "user1" cannot be dropped because some objects depend on it
DETAIL:  owner of user mapping for user1 on server foo
```

**Comment by jeeminso** [comment]
Follow up:
```
cr> drop user user1;                                                                                                                       
DROP OK, 1 row affected  (0.248 sec)
cr> CREATE USER MAPPING FOR user1 SERVER my_postgresql OPTIONS ("user2" 'trillian', password 'secret');                                    
NullPointerException[Cannot invoke "io.crate.role.Role.name()" because the return value of "io.crate.analyze.AnalyzedCreateUserMapping.user()" is null]
```

File: server/src/main/java/org/elasticsearch/common/settings/Settings.java [hunk 2]
```diff
@@ -529,7 +529,7 @@ public static Settings readSettingsFromStream(StreamInput in) throws IOException
         return builder.build();
     }
 
-    public static void writeSettingsToStream(Settings settings, StreamOutput out) throws IOException {
+    public static void writeSettingsToStream(StreamOutput out, Settings settings) throws IOException {
```
Comment (mfussenegger) [inline server/src/main/java/org/elasticsearch/common/settings/Settings.java:532]: Could you extract this change into a dedicated commit - or better PR - with an explanation why this order is better?

File: server/src/main/java/io/crate/expression/reference/StaticTableDefinition.java [hunk 3]
```diff
@@ -59,8 +60,15 @@ public StaticTableDefinition(Supplier<CompletableFuture<? extends Iterable<T>>>
...
+        this(iterable, predicate, (u, t) -> t, expressionFactories);
+    }
+
+    public StaticTableDefinition(Supplier<? extends Iterable<T>> iterable,
+                                 BiPredicate<Role, T> predicate,
+                                 BiFunction<Role, T, T> mapper,
+                                 Map<ColumnIdent, ? extends RowCollectExpressionFactory<T>> expressionFactories) {
         this.getRecords = (txnCtx, u) -> completedFuture(() -> StreamSupport.stream(iterable.get().spliterator(), false)
-            .filter(t -> u == null || predicate.test(u, t)).iterator());
+            .filter(t -> u == null || predicate.test(u, t)).map(t -> mapper.apply(u, t)).iterator());
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/reference/StaticTableDefinition.java:71]: Couldn't this be done on the call-site of the constructor instead of adding a new overload? `StaticTableDefinition` already has so many...

File: server/src/main/java/io/crate/fdw/JdbcForeignDataWrapper.java [hunk 4]
```diff
@@ -88,12 +86,12 @@ public CompletableFuture<BatchIterator<Row>> getIterator(Role currentUser,
...
-        Map<String, Object> userOptions = server.users().get(sessionSettings.userName());
+        Settings userOptions = server.users().get(sessionSettings.userName());
         if (userOptions == null) {
-            userOptions = Map.of();
+            userOptions = Settings.EMPTY;
         }
-        String user = DataTypes.STRING.implicitCast(userOptions.get("user"));
-        String password = DataTypes.STRING.implicitCast(userOptions.get("password"));
+        String user = userOptions.get("user");
+        String password = userOptions.get("password");
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/fdw/JdbcForeignDataWrapper.java:94]: this should go through the setting definition. Same as the remaining options (e.g. urlSetting)

File: server/src/main/java/io/crate/fdw/ServersMetadata.java [hunk 5]
```diff
@@ -339,6 +360,17 @@ public Iterable<UserMapping> getUserMappings() {
                 ).iterator();
     }
 
+    public Iterable<UserMappingOptionsTableInfo.UserMappingOptions> getUserMappingOptions() {
+        return () ->
+            servers.values().stream()
+                .map(server -> server)
+                .flatMap(server -> server.users.entrySet().stream()
+                    .flatMap(e -> e.getValue().getAsStructuredMap().entrySet().stream()
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/fdw/ServersMetadata.java:368]: This should probably end up using the `getAsStructuredMap` overload that takes a set of masked settings. And the masked settings need to be provided from the concrete fdw implementation.

Maybe easier to split this PR into two parts: First setting validation, and then follow up with the information schema table.

File: server/src/main/java/io/crate/planner/CreateUserMappingPlan.java [hunk 6]
```diff
@@ -38,10 +42,25 @@
 
 public class CreateUserMappingPlan implements Plan {
 
+    private static final Set<String> SUPPORTED_OPTIONS = Set.of("user", "password");
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/planner/CreateUserMappingPlan.java:45]: The available set of options depends on the concrete ForeignDataWrapper implementation. We can't hardcode this to  user/password for all of them.

This should be done in the same way as it was done for create server and create foreign table.
See https://github.com/crate/crate/blob/ddef9b04a188251a375aacef9f6a2411ff8a78ad/server/src/main/java/io/crate/planner/CreateServerPlan.java#L76-L77

Comment (jeeminso) [reply server/src/main/java/io/crate/planner/CreateUserMappingPlan.java:45]: Thank you, initially assumed that all implementations required those. Working on a separate PR https://github.com/crate/crate/pull/15606 following CreateServerPlan.

File: server/src/main/java/org/elasticsearch/common/settings/Settings.java (hunk 2 above)
Comment (jeeminso) [reply server/src/main/java/org/elasticsearch/common/settings/Settings.java:532]: Extracted to https://github.com/crate/crate/pull/15608, Thank you

File: server/src/main/java/io/crate/fdw/ServersMetadata.java (hunk 5 above)
Comment (jeeminso) [reply server/src/main/java/io/crate/fdw/ServersMetadata.java:368]: Could we instead mask all option values(not only password) without proper accesses like how Postgres handles it?

https://www.postgresql.org/docs/current/infoschema-user-mapping-options.html
>  This column will show as null unless the current user is the user being mapped, or the mapping is for PUBLIC and the current user is the server owner, or the current user is a superuser

File: server/src/main/java/io/crate/fdw/JdbcForeignDataWrapper.java (hunk 4 above)
Comment (jeeminso) [reply server/src/main/java/io/crate/fdw/JdbcForeignDataWrapper.java:93]: Once https://github.com/crate/crate/pull/15606 is merged, the settings will be available to use here.

File: docs/sql/statements/create-user-mapping.rst [hunk 7]
```diff
@@ -39,7 +39,8 @@ more details.
 
 Creating a user mapping requires ``AL`` permission on cluster level.
 
-User mappings are visible in the :ref:`user_mappings` table.
+User mappings and the options are visible in the :ref:`user_mappings` table and
+:ref:`user_mapping_options`.
```
Comment (mfussenegger) [inline docs/sql/statements/create-user-mapping.rst:43]:
```suggestion
User mappings are visible in the :ref:`user_mappings` table and the options in
:ref:`user_mapping_options`.
```

File: server/src/main/java/io/crate/fdw/ServersMetadata.java (hunk 5 above)
Comment (mfussenegger) [reply server/src/main/java/io/crate/fdw/ServersMetadata.java:368]: Not sure. I'd tend to use the masking system we have in place, to keep it consistent with sys.repositories.
The username could be visible for all and the password masked. It would also reduce the amount of custom logic required in https://github.com/crate/crate/pull/15591/files#diff-0d97afa2e4b2cf61789649f72feb9a086b2b0489f5a4a61229a53890a8deaa6cR133

Let's see what the others think.

@seut @matriv

File: server/src/main/java/io/crate/planner/CreateUserMappingPlan.java [hunk 8]
```diff
@@ -87,8 +88,9 @@ public void executeOrFail(DependencyCarrier dependencies,
             plannerContext.nodeContext(),
             subQueryResults
         ).bind(params);
-        Map<String, Object> options = createUserMapping.options().entrySet().stream()
-            .collect(Collectors.toMap(Entry::getKey, entry -> toValue.apply(entry.getValue())));
+        Map<String, String> optionsMap = createUserMapping.options().entrySet().stream()
+            .collect(Collectors.toMap(Entry::getKey, entry -> toValue.apply(entry.getValue()).toString()));
+        var options = Settings.builder().putProperties(optionsMap, s -> s).build();
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/planner/CreateUserMappingPlan.java:93]: This could break for list options if the value is converted via `.toString()` and creating a map just to then transfer it into another map in the settings builder is a waste.

Could do this like in https://github.com/crate/crate/blob/0f47b8185b6681f06f3798c5392d11646a377d46/server/src/main/java/io/crate/planner/CreateServerPlan.java#L74-L90

File: server/src/test/java/io/crate/integrationtests/ForeignDataWrapperITest.java [hunk 9]
```diff
@@ -316,4 +316,48 @@ public void test_can_use_joins_on_foreign_tables() throws Exception {
...
+        assertThat(response).hasRows("user1| crate| my_postgresql| password| NULL");
+
+        // superuser can see the pw
+        response = execute("select * from information_schema.user_mapping_options where option_name = 'password'");
+        assertThat(response).hasRows("user1| crate| my_postgresql| password| foreign-pw");
+
+        execute("drop user mapping for user1 server my_postgresql");
+        assertThat(execute("select * from information_schema.user_mapping_options where option_name = 'password'")).isEmpty();
+        execute("drop foreign table doc.remote_documents");
+        execute("drop server my_postgresql");
```
Comment (mfussenegger) [inline server/src/test/java/io/crate/integrationtests/ForeignDataWrapperITest.java:361]:
```suggestion
```

Could call the server `pg` and the users `trillian` and `arthur` to have the teardown method do this:

https://github.com/crate/crate/blob/0f47b8185b6681f06f3798c5392d11646a377d46/server/src/test/java/io/crate/integrationtests/ForeignDataWrapperITest.java#L41-L46

Would have the advantage that a test failure here doesn't cause the other tests to fail as well.

File: server/src/main/java/io/crate/fdw/ServersMetadata.java (hunk 5 above)
Comment (mfussenegger) [reply server/src/main/java/io/crate/fdw/ServersMetadata.java:368]: Actually, let's go ahead with this and get this in - we can still change it afterwards based on the feedback

**Review by mfussenegger — APPROVED** [review]

File: server/src/main/java/io/crate/planner/CreateUserMappingPlan.java (hunk 8 above)
Comment (jeeminso) [reply server/src/main/java/io/crate/planner/CreateUserMappingPlan.java]: Thanks updated
